### ISSUE TITLE
Dynamic rangesConcurrency based on Cassandra pool utilization

### ISSUE THREAD

**Issue by jnelson15** [palantir/atlasdb#1738]
@clockfort @rjullman and I just talked about the value of `rangesConcurrency` in Cassandra KVS `getRanges` calls (by default 64 threads) since some large deployments and products have lowered this to avoid exhausting pools when under heavy load. Clocks had an idea that rather than lowering the default concurrency across products, we could make it a function of load (measured by pool utilization) on the system. Tricky part will be handling large bursts of requests on a lightly-used stack so that we don't get multiple requests allocated a ton of threads, which would put us back at our original problem. So would probably want to make it a function of some set of moving averages so that it can ramp up gradually. On the other hand, we'll want it to decrease pretty quickly during times of load. Any thoughts on how we can achieve both? Basically different rates for ramp up and ramp down so that we can decrease parallelization quickly when we get bursts of requests but gradually increase parallelization as load decreases.

**Comment by jnelson15** [comment]
\+ @schlosna @tpetracca @jboreiko

**Comment by tpetracca** [comment]
Sounds like a good idea.

Also generally speaking 64 is weirdly large considering our default pool sizes and the fact that concurrency number is `getRanges` invocation specific, and not a global thing.

**Comment by clockfort** [comment]
>Also generally speaking 64 is weirdly large considering our default pool sizes and the fact that concurrency number is getRanges invocation specific, and not a global thing.

It's only way too big if all of your getRanges happen to go to a single C* host/replica set;

Normal case, uniformly distributed data / range queries, 6 node RF3 cluster:
user passes in 60 ranges --> we run all at once --> we do ~10 range requests per node

Bad case that we've seen recently, range reads all hitting a single partition, 6 node RF3 cluster:
user passes in 60 ranges --> we run all at once --> we do ~20 range requests per node

Us implementing the logic outlined above would also let us catch the "bad" case and handle it nicely, because we'd be able to see in that part of the code which nodes we're going to send the queries to, and could prevent ourselves from sending too many in parallel to a small number of nodes.

**Comment by rhero** [comment]
@jnelson15 yeah we'll probably want to do this. Are you already working on something? We have quite a few jmh benchmarks for getRanges, but I doubt they cover the "bad" and "normal" cases that Clockfort highlighted (since our tests are three node RF3). We may want to do some more one-off benchmarking to see the total effect.

**Comment by jnelson15** [comment]
No haven't started anything yet. Do you guys have any resources to help with this? If not I can probably find a little time later this week.

**Comment by jnelson15** [comment]
Closing in favor of https://github.com/palantir/atlasdb/issues/1706. See comment https://github.com/palantir/atlasdb/issues/1706#issuecomment-293950998 for more info.
